### ISSUE TITLE
@external directive is not properly licensed by a composite @key directive

### ISSUE THREAD

**Issue by adimit** [apollographql/federation#1353]
Repro-Repository: https://github.com/adimit/apollo-federation-bug-demo/ (This is based on the `federation-demo`. Some code is left over, but only the `gql`-schemas are relevant here.) I'm not sure which versions are affected, but versions 0.16.5 and 0.42.0 are definitely affected.

Use-Case: we have a type `BaseType` with a composite `@key` directive. One of the components of this `@key` is an entity, but the other one is not. I want to extend this base type using my complex `@key` directive.
```graphql
  type User @key(fields: "id") {
    id: ID!
  }

  type Complex {
    subproperty: String
  }

  type BaseType @key(fields: "user { id } dependency { subproperty }") {
    user: User!
    dependency: Complex!
  }
```

So all instances of `BaseType` are unique *for a certain combination of `User.id` and `Complex.subproperty`*.

I want to extend `BaseType` in a different service, and index it using this `@key`-directive. However, this fails:

```graphql
  type User @extends @key(fields: "id") {
    id: ID! @external
  }

  type Complex @extends {
    # ERROR: subproperty can't be external without @key or @requires!
    subproperty: String @external
  }

  # But subproperty is in a @key directive here! Apollo federation doesn't recognise this
  type BaseType @extends @key(fields: "user {id} dependency {subproperty}") {
    user: User! @external
    dependency: Complex! @external
  }
```

The error is:

```
This data graph is missing a valid configuration. [products] Complex.subproperty -> is marked as @external but is not used by a @requires, @key, or @provides directive.
```

I would expect this to work: `Complex.subproperty` *is* in a `@key` directive, namely in the `@key` directive for `BaseType`. This way I cannot extend `BaseType` at all without making `Complex.subproperty` a `@key`—which is not something I want to (or necessarily can) do.

There is a workaround, I could declare a dummy `@requires` on `BaseType`:

```graphql
  type BaseType @extends @key(fields: "user {id} dependency {subproperty}") {
    user: User! @external
    dependency: Complex! @external
    # A requires directive here will work.
    workaround: String! @requires(fields: "dependency {subproperty}")
  }
```

This `@requires` seems to license my use of `@external` on `subproperty`, but I don't understand why the `@key` directive didn't already take care of that.

**Comment by pcmanus** [comment]
I agree it's a bug with the validation for externals. Looking [at the code](https://github.com/apollographql/federation/blob/version-0.x/federation-js/src/composition/validate/postComposition/externalUnused.ts#L46), given an external field like `subproperty`, the validation only look at keys on the parent type (so `Complex` here), thus missing the one on `BaseType`, and besides, even for keys it looks at, it only consider the top-level fields for the key.

The bright news is that the upcoming federation 2 (currently in alpha) fixes this. The less bright news is that the validation code on `version-0.x` confuses me a bit and I'm not sure how to fix it there easily. Maybe @trevor-scheer knows a simple way to fix this?
